src/tools/clippy/clippy_utils/src/hir_utils.rs

```diff
@@ -518,8 +518,8 @@ impl HirEqInterExpr<'_, '_, '_> {
         }
     }
 
-    fn eq_type_binding(&mut self, left: &TypeBinding<'_>, right: &TypeBinding<'_>) -> bool {
-        left.ident.name == right.ident.name && self.eq_ty(left.ty(), right.ty())
+    fn eq_assoc_type_binding(&mut self, left: &AssocItemConstraint<'_>, right: &AssocItemConstraint<'_>) -> bool {
+        left.ident.name == right.ident.name && self.eq_ty(left.ty().unwrap(), right.ty().unwrap())
```

Comment:
Why are those unwraps safe to do?

Reply:
Clippy only ever calls `eq_assoc_type_binding` for parenthesized generic args (`Trait(Ty0, Ty1, ...) -> RetTy`) whose lowered form is guaranteed to contain only a single associated type binding, namely for the associated type `Output`).

I'm not saying that those unwraps are great, they certainly aren't. However, this change is entirely semantics-preserving as `TypeBinding::ty` on master panics if the assoc item constraint isn't an assoc ty equality constraint. I've made method `ty` non-panicking in this PR returning an `Option` instead, hence the need to unwrap here.

Reply:
I see. Can you leave a comment here explaining this or swap the `unwrap` to an `expect` (or both), please?